accounts/pkg/storage/cs3.go

```diff
@@ -0,0 +1,293 @@
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	"github.com/cs3org/reva/pkg/rgrpc/todo/pool"
+	"github.com/cs3org/reva/pkg/token"
+	"github.com/cs3org/reva/pkg/token/manager/jwt"
+	"github.com/owncloud/ocis/accounts/pkg/config"
+	"github.com/owncloud/ocis/accounts/pkg/proto/v0"
+	"google.golang.org/grpc/metadata"
+)
+
+// CS3Repo provides a cs3 implementation of the Repo interface
+type CS3Repo struct {
+	serviceID       string
+	cfg             *config.Config
+	tm              token.Manager
+	storageProvider provider.ProviderAPIClient
+	dataProvider    dataProviderClient // Used to create and download data via http, bypassing reva upload protocol
+}
+
+// NewCS3Repo creates a new cs3 repo
+func NewCS3Repo(serviceID string, cfg *config.Config) (Repo, error) {
+	tokenManager, err := jwt.New(map[string]interface{}{
+		"secret": cfg.TokenManager.JWTSecret,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := pool.GetStorageProviderServiceClient(cfg.Repo.CS3.ProviderAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	return CS3Repo{
+		serviceID:       serviceID,
+		cfg:             cfg,
+		tm:              tokenManager,
+		storageProvider: client,
+		dataProvider: dataProviderClient{
+			client: http.Client{
+				Transport: http.DefaultTransport,
+			},
+		},
+	}, nil
+}
+
+// WriteAccount writes an account via cs3 and modifies the provided account (e.g. with a generated id).
+func (r CS3Repo) WriteAccount(ctx context.Context, a *proto.Account) (err error) {
+	t, err := r.authenticate(ctx)
+	if err != nil {
+		return err
+	}
+
+	ctx = metadata.AppendToOutgoingContext(ctx, token.TokenHeader, t)
+	if err := r.makeRootDirIfNotExist(ctx, accountsFolder); err != nil {
+		return err
+	}
+
+	var by []byte
+	if by, err = json.Marshal(a); err != nil {
+		return err
+	}
+
+	_, err = r.dataProvider.put(r.accountURL(a.Id), bytes.NewReader(by), t)
+	return err
+}
+
+// LoadAccount loads an account via cs3 by id and writes it to the provided account
+func (r CS3Repo) LoadAccount(ctx context.Context, id string, a *proto.Account) (err error) {
+	t, err := r.authenticate(ctx)
+	if err != nil {
+		return err
+	}
+
+	resp, err := r.dataProvider.get(r.accountURL(id), t)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode == http.StatusNotFound {
+		return &notFoundErr{"account", id}
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if err := json.Unmarshal(b, &a); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// DeleteAccount deletes an account via cs3 by id
+func (r CS3Repo) DeleteAccount(ctx context.Context, id string) (err error) {
+	t, err := r.authenticate(ctx)
+	if err != nil {
+		return err
+	}
+
+	ctx = metadata.AppendToOutgoingContext(ctx, token.TokenHeader, t)
+
+	resp, err := r.storageProvider.Delete(ctx, &provider.DeleteRequest{
+		Ref: &provider.Reference{
+			Spec: &provider.Reference_Path{Path: fmt.Sprintf("/meta/%s/%s", accountsFolder, id)},
```

Comment:
use `path` here for more robust handling. Even if there is deffensive code in `accountsFolder` or `id`, if one of them does not exist we can have multiple consecutive slashes, while using `path` that's taken care of:

> Join joins any number of path elements into a single path, separating them with slashes. Empty elements are ignored. The result is Cleaned. However, if the argument list is empty or all its elements are empty, Join returns an empty string.